api/src/repositories/alert-repository.ts

```diff
@@ -0,0 +1,228 @@
+   *
+   * @return {*}  {Knex.QueryBuilder}
+   * @memberof AlertRepository
+   */
+  _getAlertBaseQuery(): Knex.QueryBuilder {
+    const knex = getKnex();
+
+    return knex
+      .select(
+        'alert.alert_id',
+        'alert.name',
+        'alert.message',
+        'alert.alert_type_id',
+        'alert.data',
+        'alert.severity',
+        'alert.record_end_date',
+        knex.raw(`
+          CASE
+            WHEN alert.record_end_date < NOW() THEN 'expired'
+            ELSE 'active'
+          END AS status
+        `)
+      )
+      .from('alert')
+      .orderBy('alert.create_date', 'DESC');
+  }
+
+  /**
+   * Get alert records with optional filters applied
+   *
+   * @param {IAlertFilterObject} filterObject
+   * @return {*}  {Promise<IAlert[]>}
+   * @memberof AlertRepository
+   */
+  async getAlerts(filterObject: IAlertFilterObject): Promise<IAlert[]> {
+    const queryBuilder = this._getAlertBaseQuery();
+
+    if (filterObject.expiresAfter) {
+      queryBuilder.where((qb) => {
+        qb.whereRaw(`alert.record_end_date >= ?`, [filterObject.expiresAfter]).orWhereNull('alert.record_end_date');
+      });
+    }
+
+    if (filterObject.expiresBefore) {
+      queryBuilder.where((qb) => {
+        qb.whereRaw(`alert.record_end_date < ?`, [filterObject.expiresBefore]);
+      });
+    }
+
+    if (filterObject.types && filterObject.types.length > 0) {
+      queryBuilder
+        .join('alert_type as at', 'at.alert_type_id', 'alert.alert_type_id')
+        .whereRaw('lower(at.name) = ANY(?)', [filterObject.types.map((type) => type.toLowerCase())]);
+    }
+
+    const response = await this.connection.knex(queryBuilder, IAlert);
+
+    return response.rows;
+  }
+
+  /**
+   * Get a specific alert by its Id
+   *
+   * @param {number} alertId
+   * @return {*}  {Promise<IAlert>}
+   * @memberof AlertRepository
+   */
+  async getAlertById(alertId: number): Promise<IAlert> {
+    const queryBuilder = this._getAlertBaseQuery();
+
+    queryBuilder.where('alert_id', alertId);
+
+    const response = await this.connection.knex(queryBuilder, IAlert);
+
+    return response.rows[0];
+  }
+
+  /**
+   * Update system alert.
+   *
+   * @param {IAlertUpdateObject} alert
+   * @return {*} Promise<number>
+   * @memberof AlertRepository
+   */
+  async updateAlert(alert: IAlertUpdateObject): Promise<number> {
+    const sqlStatement = SQL`
+      UPDATE alert
+      SET
+        name = ${alert.name},
+        message = ${alert.message},
+        alert_type_id = ${alert.alert_type_id},
+        severity = ${alert.severity},
+        data = ${JSON.stringify(alert.data)}::json,
+        record_end_date = ${alert.record_end_date}
+      WHERE
+        alert_id = ${alert.alert_id}
+      RETURNING alert_id
+      ;
+      `;
+
+    const response = await this.connection.sql(sqlStatement);
+
+    const result = response?.rows?.[0] ?? null;
+
+    if (!result) {
```

Comment:
Its slightly better to check `response.rowCount`, which will always be the number of rows affected. Why that matters is because if we make an update query, and forget a where clause, it might be that it updates more than 1 record.

And so we really want to check if `response.rowCount === 1`. Because less than 1 or more than 1 is equally bad (in this case).

Reply:
You'll see some checks like that in SIMS, and the message will be:
`''response.rowCount was !== 1, expected rowCount === 1''`

Reply:
Thanks, that makes sense